src/loader/bgw_message_queue.c

```diff
@@ -0,0 +1,210 @@
+	tsbgwMessage buffer[TSBGW_MAX_MESSAGES];
+}			tsbgwMessageQueue;
+
+
+
+/*
+ * Attach to the message queue, which just returns the static variable we assigned in normal cases,
+ * assign it are in a possible restart case, which only happens during the shmem startup hook
+ * in the case where we do not find the shared memory segment we must initialize it. (it's also why we need the lock)
+ */
+
+static tsbgwMessageQueue * tsbgw_message_queue_attach(bool possible_restart)
+{
+	static tsbgwMessageQueue * tsbgw_mq = NULL;
+	bool		found;
+
+	/* reset in case this is a restart within the postmaster */
+	if (possible_restart)
+		tsbgw_mq = NULL;
+	if (tsbgw_mq == NULL)
+	{
+		LWLockAcquire(AddinShmemInitLock, LW_EXCLUSIVE);
+		tsbgw_mq = ShmemInitStruct(TSBGW_MESSAGE_QUEUE_NAME, sizeof(tsbgwMessageQueue), &found);
+		if (!found)
+		{
+			memset(tsbgw_mq, 0, sizeof(tsbgwMessageQueue));
+			tsbgw_mq->lock = &(GetNamedLWLockTranche(TSBGW_MQ_TRANCHE_NAME))->lock;
+		}
+		LWLockRelease(AddinShmemInitLock);
+	}
+	return tsbgw_mq;
+}
+
+
+/* this gets called when shared memory is initialized in a backend (shmem_startup_hook) */
+extern void
+tsbgw_message_queue_shmem_startup(void)
+{
+	tsbgw_message_queue_attach(true);
+}
+
+/* this is called in the loader during server startup to allocate a shared memory segment*/
+extern void
+tsbgw_message_queue_alloc(void)
+{
+	RequestAddinShmemSpace(sizeof(tsbgwMessageQueue));
+	RequestNamedLWLockTranche(TSBGW_MQ_TRANCHE_NAME, 1);
+}
+
+
+/* Notes on managing the queue/locking:
+ * We decided that for this application, simplicity of locking scheme was more important than
+ * being very good about concurrency as the frequency of these messages will be low and the
+ * number of messages on this queue should be low, given that they mostly happen when we update
+ * the extension. Therefore we decided to simply take an exclusive lock whenever we were modifying
+ * anything in the shared memory segment to avoid collisions.
+ */
+
+/* Add a message to the queue - we can do this if the queue is not full */
+static bool
+tsbgw_message_queue_add(tsbgwMessageQueue * queue, tsbgwMessage * message)
+{
+	bool		message_sent = false;
+	pid_t		reader_pid = 0;
+
+	LWLockAcquire(queue->lock, LW_EXCLUSIVE);
+	if (queue->num_elements < TSBGW_MAX_MESSAGES)
+	{
+		memcpy(&queue->buffer[(queue->read_upto + queue->num_elements) % TSBGW_MAX_MESSAGES], message, sizeof(tsbgwMessage));
+		queue->num_elements++;
+		message_sent = true;
+		reader_pid = queue->reader_pid;
+	}
+	LWLockRelease(queue->lock);
+
+	if (reader_pid != 0)		/* set latch on the reader as long as one is
+								 * defined, even if queue is full, then we def
+								 * need to read messages */
+		SetLatch(&BackendPidGetProc(reader_pid)->procLatch);
+
+	return message_sent;
+}
+
+static tsbgwMessage * tsbgw_message_queue_remove(tsbgwMessageQueue * queue)
+{
+	tsbgwMessage *message = NULL;
+
+	LWLockAcquire(queue->lock, LW_EXCLUSIVE);
+	if (queue->reader_pid == 0)
```

Comment:
There is a postgres #define for InvalidPid you should use that instead of 0. Please replace everywhere.